test/conftest.py

```diff
@@ -291,56 +286,186 @@ def pytest_collection_modifyitems(items: List[Any]) -> None:
 
 
 class StepcurrentPlugin:
-    # Modified fromo _pytest/stepwise.py in order to save the currently running
-    # test instead of the last failed test
+    """
+    This is meant to work in with test/run_test.py to ensure that every test is
+    run, gets retries in new subprocesses, and creates xml.  To do this, it
+    keeps track of each test's status in the cache.  Normal tests are run
+    together in the same process, but when a test fails, it needs to be run
+    singly.  This class and run_test.py have logic to ensure that this happens,
+    especially for unusual cases like segfaults where the process exits
+    immediately and doesn't produce xml for any of the tests, and tests that
+    cause the process to fail at exit which creates xml that believes the test
+    is successful.
+
+    If a test fails normally, it gets run singly in a new subprocess.  If a test
+    segfaults, all tests prior to the segfault get rerun to generate xml, and
+    the segfaulting test gets run singly in a new subprocess.  If a process
+    exist with non 0 exit code but no tests seem to fail, then all the test will
+    be rerun singly to narrow down which test caused the failure.
+
+    Cache file contents:
+    pytest_previous_status: None on the first run, and then is either an exit
+    code or "no xml"
+    to_run: all tests that have still not created xml
+    prev_run: tests that are expected to run during this run.  Poorly named in
+    this context, but makes sense from test/run_test.py's POV.  If a test fails,
+    then it is possibly that not all tests in this actually ran. This should be
+    a prefix of to_run
+    already_ran: tests that have created xml
+
+    Test that are still to be run have 4 possible statuses, which are in
+    test/run_test.py and are: cont, s, r2, and r1, abbreviated c, s, 2, and 1 in
+    the table
+
+    cont: Test can be run in same process as other tests, this is the default status
+    s: Test should be run singly
+    r2: Test should be run singly, and has 2 retries left
+    r1: Test should be run singly, and has 1 retry left
+
+    At the start of the run, all tests get the status c. This plugin will either
+    run all tests with status c, or run a single test with the status s, 2, or
+    1. If a test is known to fail, then the status of the test is changed to 2
+    or 1 depending on how many retires it still has left.
+
+    Examples failure causes with 3 tests:
+    +-------------+--------+----------+
+    | already_ran | to_run | prev_run |
+    +-------------+--------+----------+
```

Comment:
nit: A explainer about how to parse the chart below.  Those charts really help btw


```suggestion
    Examples failure causes with 3 tests:
    +-------------+--------+----------+
    | already_ran | to_run | prev_run |
    +-------------+--------+----------+
    
    Each row in the images below should the state after successive runs of the given scenario
```